### PR TITLE
SQL: tag filter for the 2nd table removes all rows in JOIN #3756

### PR DESCRIPTION
**Pull request by RawrLion**
Added tests for #3756

### CODE REVIEW

File: src/test/java/com/axibase/tsd/api/method/sql/clause/join/SqlClauseJoinWithTagFilter.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+                "SELECT t1.value, t2.value FROM '%s' t1 JOIN '%s' t2 WHERE t1.entity IS NOT NULL " +
+                        "AND t2.entity IS NOT NULL AND t1.entity = '%s' AND t2.entity = '%s'",
+                TEST_METRIC1_NAME, TEST_METRIC2_NAME, TEST_ENTITY_NAME, TEST_ENTITY_NAME
+        );
+
+        String[][] expectedRows = {
+                {"1", "2"}
+        };
+
+        String assertMessage = "JOIN with Entity filter gives wrong result";
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/sql/clause/join/SqlClauseJoinWithTagFilter.java]: you can inline this var and others like that if you want

File: src/test/java/com/axibase/tsd/api/method/sql/clause/join/SqlClauseJoinWithTagFilter.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    /**
+     * #3756
+     */
+    @Test
+    public void testJoinWithTagFilter() {
+        String sqlQuery = String.format(
+                "SELECT t1.value, t2.value FROM '%s' t1 JOIN '%s' t2 WHERE t1.tags.tag IS NOT NULL " +
+                        "AND t2.tags.tag IS NOT NULL AND t1.tags.tag = \'123\' AND t2.tags.tag = \'123\'",
```
Comment (VeselovAlex) [inline src/test/java/com/axibase/tsd/api/method/sql/clause/join/SqlClauseJoinWithTagFilter.java]: Do you really need escaping?

File: src/test/java/com/axibase/tsd/api/method/sql/clause/join/SqlClauseJoinWithTagFilter.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+import org.testng.annotations.Test;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static com.axibase.tsd.api.util.Util.TestNames.entity;
+import static com.axibase.tsd.api.util.Util.TestNames.metric;
+
+public class SqlClauseJoinWithTagFilter extends SqlTest {
```
Comment (VeselovAlex) [inline src/test/java/com/axibase/tsd/api/method/sql/clause/join/SqlClauseJoinWithTagFilter.java]: Class name is redundant, IMHO

File: src/test/java/com/axibase/tsd/api/method/sql/clause/join/SqlClauseJoinWithTagFilter.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    /**
+     * #3756
+     */
+    @Test
+    public void testJoinUsingEntityWithTagFilter() {
+        String sqlQuery = String.format(
+                "SELECT t1.value, t2.value, t3.value FROM '%s' t1 JOIN USING ENTITY '%s' t2 JOIN USING ENTITY '%s' t3 " +
+                        "WHERE t1.tags.tag IS NOT NULL AND t2.tags.tag IS NOT NULL " +
```
Comment (VeselovAlex) [inline src/test/java/com/axibase/tsd/api/method/sql/clause/join/SqlClauseJoinWithTagFilter.java]:
```
sql t1.tags.tag IS NOT NULL ... AND t1.tags.tag = \'123\' 
````
Do we need IS NOT NULL check?

File: src/test/java/com/axibase/tsd/api/method/sql/clause/join/SqlClauseJoinWithTagFilter.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+
+    /**
+     * #3756
+     */
+    @Test
+    public void testJoinWithEntityFilter() {
+        String sqlQuery = String.format(
+                "SELECT t1.value, t2.value FROM '%s' t1 JOIN '%s' t2 WHERE t1.entity IS NOT NULL " +
+                        "AND t2.entity IS NOT NULL AND t1.entity = '%s' AND t2.entity = '%s'",
```
Comment (VeselovAlex) [inline src/test/java/com/axibase/tsd/api/method/sql/clause/join/SqlClauseJoinWithTagFilter.java]: Format query, e.g
```java
"SELECT t1.value, t2.value " + 
"FROM '%s' t1 JOIN '%s' t2 " + 
"WHERE t1.entity IS NOT NULL AND t2.entity IS NOT NULL AND " +
     "t1.entity = '%s' AND t2.entity = '%s'",
```

File: src/test/java/com/axibase/tsd/api/method/sql/clause/join/JoinWithTagOrEntityFilter.java [hunk 6]
```diff
@@ -75,9 +76,9 @@ public void testJoinWithEntityFilter() {
     @Test
     public void testJoinUsingEntityWithEntityFilter() {
         String sqlQuery = String.format(
-                "SELECT t1.value, t2.value, t3.value FROM '%s' t1 JOIN USING ENTITY '%s' t2 JOIN USING ENTITY '%s' t3 " +
-                        "WHERE t1.entity IS NOT NULL AND t2.entity IS NOT NULL " +
-                        "AND t3.entity IS NOT NULL AND t1.entity = '%s' " +
+                "SELECT t1.value, t2.value, t3.value " +
+                "FROM '%s' t1 JOIN USING ENTITY '%s' t2 JOIN USING ENTITY '%s' t3 " +
+                "WHERE t1.entity IS NOT NULL AND t2.entity IS NOT NULL AND t3.entity IS NOT NULL AND t1.entity = '%s' " +
```
Comment (VeselovAlex) [inline src/test/java/com/axibase/tsd/api/method/sql/clause/join/JoinWithTagOrEntityFilter.java]: May be better to place not-null checks to one line, and equal checks to other line?

File: src/test/java/com/axibase/tsd/api/method/sql/clause/join/JoinWithTagOrEntityFilter.java [hunk 7]
```diff
@@ -75,9 +76,9 @@ public void testJoinWithEntityFilter() {
...
                         "AND t2.entity = '%s' AND t3.entity = '%s'",
                 TEST_METRIC1_NAME, TEST_METRIC2_NAME, TEST_METRIC3_NAME, TEST_ENTITY_NAME, TEST_ENTITY_NAME, TEST_ENTITY_NAME
```
Comment (VeselovAlex) [inline src/test/java/com/axibase/tsd/api/method/sql/clause/join/JoinWithTagOrEntityFilter.java]: May be we should useargument indices, like here:
```java
Calendar c = ...;
String s = String.format("Duke's Birthday: %1$tm %1$te,%1$tY", c);
```

File: src/test/java/com/axibase/tsd/api/method/sql/clause/join/JoinWithTagOrEntityFilter.java [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+        for (int i = 0; i < metricNames.length; i++) {
+            String metricName = metricNames[i];
+            Registry.Metric.register(metricName);
+
+            Series series = new Series();
+            series.setEntity(TEST_ENTITY_NAME);
+            series.setMetric(metricName);
+
+            series.setData(Collections.singletonList(
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/sql/clause/join/JoinWithTagOrEntityFilter.java]: use addData instead

File: src/test/java/com/axibase/tsd/api/method/sql/clause/join/JoinWithTagOrEntityFilter.java [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+public class JoinWithTagOrEntityFilter extends SqlTest {
+    private static final String TEST_METRIC1_NAME = metric();
+    private static final String TEST_METRIC2_NAME = metric();
+    private static final String TEST_METRIC3_NAME = metric();
+    private static final String TEST_ENTITY_NAME = entity();
+
+    @BeforeClass
+    public static void prepareData() throws Exception {
+        List<Series> seriesList = new ArrayList<>();
```
Comment (unrealwork) [inline src/test/java/com/axibase/tsd/api/method/sql/clause/join/JoinWithTagOrEntityFilter.java]: What is the reason of the field?

**Pull request merged**